apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs

```diff
@@ -0,0 +1,334 @@
+            request.Transaction = new TransactionSelector { Id = TransactionId };
+
+            var settings = CreateSettings(Client.Settings.PartitionQuerySettings, timeoutSeconds, cancellationToken);
+            return RecordSuccessAndExpiredSessions(Client.PartitionQueryAsync(request, settings));
+        }
+
+        /// <summary>
+        /// Creates a <see cref="ReliableStreamReader"/> for the given request
+        /// </summary>
+        /// <param name="request">The query request. Must not be null. The request will be modified with session and transaction details
+        /// from this object. If this object's <see cref="TransactionId"/> is null, the request's transaction is not modified.</param>
+        /// <param name="timeoutSeconds">The timeout for this RPC, in seconds.</param>
+        /// <returns>A <see cref="ReliableStreamReader"/> for the streaming SQL request.</returns>
+        public ReliableStreamReader ExecuteSqlStreamReader(ExecuteSqlRequest request, int timeoutSeconds)
+        {
+            CheckNotDisposed();
+            GaxPreconditions.CheckNotNull(request, nameof(request));
+            if (TransactionId != null)
+            {
+                request.Transaction = new TransactionSelector { Id = TransactionId };
+            }
+            return Client.GetSqlStreamReader(request, _session, timeoutSeconds);
+        }
+
+        /// <summary>
+        /// Executes an ExecuteSql RPC asynchronously.
+        /// </summary>
+        /// <param name="request">The query request. Must not be null. The request will be modified with session and transaction details
+        /// from this object. If this object's <see cref="TransactionId"/> is null, the request's transaction is not modified.</param>
+        /// <param name="timeoutSeconds">The timeout for this RPC, in seconds.</param>
+        /// <param name="cancellationToken">An optional token for canceling the call.</param>
+        /// <returns>A task representing the asynchronous operation. When the task completes, the result is the response from the RPC.</returns>
+        public Task<ResultSet> ExecuteSqlAsync(ExecuteSqlRequest request, int timeoutSeconds, CancellationToken cancellationToken)
+        {
+            CheckNotDisposed();
+            GaxPreconditions.CheckNotNull(request, nameof(request));
+            request.SessionAsSessionName = SessionName;
+            if (TransactionId != null)
+            {
+                request.Transaction = new TransactionSelector { Id = TransactionId };
+            }
+            var settings = CreateSettings(Client.Settings.ExecuteSqlSettings, timeoutSeconds, cancellationToken);
+            return RecordSuccessAndExpiredSessions(Client.ExecuteSqlAsync(request, settings));
+        }
+
+        /// <summary>
+        /// Executes a BeginTransaction RPC asynchronously.
+        /// </summary>
+        /// <remarks>
+        /// This method does not affect <see cref="TransactionId"/> of this object.
```

Comment:
Is this like a nested transaction then?

Reply:
I like the separation of RPC and management.

But this method is `public` and `WithTransaction` is `internal`, I think they both need to be `public` and this one needs more documentation, or they both need to be `internal` and only be called from `SessionPool` (I prefer both `internal`).
Right now nothing, not even docs, is preventing client code from calling `BeginTransactionAsync` and then `Commit` and that, if the original transaction was stomped, will definetely failed.

Reply:
Either way works better than how it is now, I prefer `internal` but I'm really fine with either, really.